File: src/origins.go
package main

import (
	"net/http"
	"strings"
)

// Documents are served from a different hostname than the reader, so an
// uploaded file is a stranger to the page framing it. The browser then refuses
// it any access to the reader's DOM or its session, which is what lets the
// document run its own scripts safely: charts, maps, anything.
//
// A different port would not do. Cookies ignore ports, so a document on
// another port could still make requests carrying the reader's session. It has
// to be a different host.
//
// The names are derived rather than configured: whatever host the reader is
// on, documents live on "docs." in front of it. On a real domain that is one
// DNS record and one certificate; in development, browsers resolve anything
// ending in .localhost by themselves, so it needs no setup at all.

const docsPrefix = "docs."

// isDocsHost says whether this request arrived on the document hostname.
func isDocsHost(r *http.Request) bool {
	return strings.HasPrefix(strings.ToLower(r.Host), docsPrefix)
}

// docsHost is where documents for this deployment live.
func docsHost(host string) string {
	if strings.HasPrefix(strings.ToLower(host), docsPrefix) {
		return host
	}
	return docsPrefix + host
}

// readerHost is the inverse: the reader that owns a document hostname.
func readerHost(host string) string {
	return strings.TrimPrefix(host, docsPrefix)
}

func requestScheme(r *http.Request) string {
	if r.TLS != nil || r.Header.Get("X-Forwarded-Proto") == "https" {
		return "https"
	}
	return "http"
}

// docsOrigin is the origin the reader frames documents from, and the only
// origin it accepts postMessage traffic from.
func docsOrigin(r *http.Request) string {
	return requestScheme(r) + "://" + docsHost(r.Host)
}

func readerOrigin(r *http.Request) string {
	return requestScheme(r) + "://" + readerHost(r.Host)
}

// crossSiteRefused applies rule A to a state-changing route a browser can
// reach with cookies attached: docs.<host> is same-site with the reader, so
// SameSite cookies alone do not stop a hostile document from posting here.
// A bearer token (the CLI) skips all of this -- it is never attached to a
// request automatically, so a hostile page cannot forge one. Otherwise all
// three must hold: any Origin header sent must be this reader's own origin,
// any Sec-Fetch-Site header must say the request was not cross-site, and a
// custom header must be present, which a browser cannot attach to a
// cross-origin request without a CORS preflight that is never granted.
func crossSiteRefused(r *http.Request) bool {
	if strings.HasPrefix(r.Header.Get("Authorization"), "Bearer ") {
		return false
	}
	if origin := r.Header.Get("Origin"); origin != "" && origin != readerOrigin(r) {
		return true
	}
	if site := r.Header.Get("Sec-Fetch-Site"); site != "" && site != "same-origin" && site != "none" {
		return true
	}
	return r.Header.Get("X-Komodoc-Client") == ""
}

// wsOriginRefused is rule A's WebSocket variant: browsers always send Origin
// on a WebSocket handshake and cannot be made to skip it or to attach a
// custom header, so the custom-header check does not apply here -- an absent
// Origin is not itself suspicious, but a foreign one is refused.
func wsOriginRefused(r *http.Request) bool {
	origin := r.Header.Get("Origin")
	return origin != "" && origin != readerOrigin(r)
}

// crossSiteRefusal is the JSON body every refusal under rule A answers with.
func crossSiteRefusal() map[string]any {
	return map[string]any{"error": "cross-site request refused"}
}

File: komodoc/src/origins.rs
//! Documents are served from a different hostname than the reader, so an
//! uploaded file is a stranger to the page framing it. The browser then refuses
//! it any access to the reader's DOM or its session, which is what lets the
//! document run its own scripts safely: charts, maps, anything.
//!
//! A different port would not do. Cookies ignore ports, so a document on
//! another port could still make requests carrying the reader's session. It
//! has to be a different host.
//!
//! The names are derived rather than configured: whatever host the reader is
//! on, documents live on "docs." in front of it. On a real domain that is one
//! DNS record and one certificate; in development, browsers resolve anything
//! ending in .localhost by themselves, so it needs no setup at all.

use axum::http::HeaderMap;

pub const DOCS_PREFIX: &str = "docs.";

/// What a request tells about where it arrived: the host it was addressed to
/// and the scheme it arrived over, read once from the headers so every check
/// below sees the same answer.
#[derive(Clone, Debug)]
pub struct Arrival {
    pub host: String,
    pub scheme: &'static str,
}

impl Arrival {
    pub fn from_headers(headers: &HeaderMap) -> Arrival {
        let host = header(headers, "host").unwrap_or_default();
        let scheme = if header(headers, "x-forwarded-proto").as_deref() == Some("https") {
            "https"
        } else {
            "http"
        };
        Arrival { host, scheme }
    }

    /// Whether this request arrived on the document hostname.
    pub fn is_docs_host(&self) -> bool {
        self.host.to_lowercase().starts_with(DOCS_PREFIX)
    }

    /// The origin the reader frames documents from, and the only origin it
    /// accepts postMessage traffic from.
    pub fn docs_origin(&self) -> String {
        format!("{}://{}", self.scheme, docs_host(&self.host))
    }

    pub fn reader_origin(&self) -> String {
        format!("{}://{}", self.scheme, reader_host(&self.host))
    }

    pub fn is_https(&self) -> bool {
        self.scheme == "https"
    }

    pub fn callback_url(&self) -> String {
        format!("{}://{}/auth/callback", self.scheme, self.host)
    }
}

pub fn header(headers: &HeaderMap, name: &str) -> Option<String> {
    headers
        .get(name)
        .and_then(|value| value.to_str().ok())
        .map(str::to_string)
}

/// Where documents for this deployment live.
pub fn docs_host(host: &str) -> String {
    if host.to_lowercase().starts_with(DOCS_PREFIX) {
        host.to_string()
    } else {
        format!("{DOCS_PREFIX}{host}")
    }
}

/// The inverse: the reader that owns a document hostname.
pub fn reader_host(host: &str) -> String {
    host.strip_prefix(DOCS_PREFIX).unwrap_or(host).to_string()
}

/// Rule A, for a state-changing route a browser can reach with cookies
/// attached: docs.<host> is same-site with the reader, so SameSite cookies
/// alone do not stop a hostile document from posting here. A bearer token
/// (the CLI) skips all of this -- it is never attached to a request
/// automatically, so a hostile page cannot forge one. Otherwise all three must
/// hold: any Origin header sent must be this reader's own origin, any
/// Sec-Fetch-Site header must say the request was not cross-site, and a custom
/// header must be present, which a browser cannot attach to a cross-origin
/// request without a CORS preflight that is never granted.
pub fn cross_site_refused(headers: &HeaderMap, arrival: &Arrival) -> bool {
    if header(headers, "authorization").is_some_and(|value| value.starts_with("Bearer ")) {
        return false;
    }
    if let Some(origin) = header(headers, "origin") {
        if !origin.is_empty() && origin != arrival.reader_origin() {
            return true;
        }
    }
    if let Some(site) = header(headers, "sec-fetch-site") {
        if !site.is_empty() && site != "same-origin" && site != "none" {
            return true;
        }
    }
    header(headers, "x-komodoc-client").is_none_or(|value| value.is_empty())
}

/// Rule A's WebSocket variant: browsers always send Origin on a WebSocket
/// handshake and cannot be made to skip it or to attach a custom header, so
/// the custom-header check does not apply here -- an absent Origin is not
/// itself suspicious, but a foreign one is refused.
pub fn ws_origin_refused(headers: &HeaderMap, arrival: &Arrival) -> bool {
    match header(headers, "origin") {
        Some(origin) => !origin.is_empty() && origin != arrival.reader_origin(),
        None => false,
    }
}

/// The JSON body every refusal under rule A answers with.
pub fn cross_site_refusal() -> serde_json::Value {
    serde_json::json!({"error": "cross-site request refused"})
}
